src/Microsoft.VisualStudio.Editor.Razor/BraceSmartIndenter.cs

```diff
@@ -0,0 +1,266 @@
+            edit.Insert(insertLocation, indent);
+            edit.Apply();
+        }
+
+        // Internal for testing
+        internal static void RestoreCaretTo(int caretPosition, ITextView textView)
+        {
+            var currentSnapshotPoint = new SnapshotPoint(textView.TextBuffer.CurrentSnapshot, caretPosition);
+            textView.Caret.MoveTo(currentSnapshotPoint);
+        }
+
+        // Internal for testing
+        internal static bool TryCreateIndentationContext(int changePosition, int changeLength, string finalText, VisualStudioDocumentTracker documentTracker, out BraceIndentationContext context)
+        {
+            var focusedTextView = documentTracker.GetFocusedTextView();
+            if (focusedTextView != null && ParserHelpers.IsNewLine(finalText))
+            {
+                var currentSnapshot = documentTracker.TextBuffer.CurrentSnapshot;
+                var preChangeLineSnapshot = currentSnapshot.GetLineFromPosition(changePosition);
+
+                // Handle the case where the \n comes through separately from the \r and the position
+                // on the line is beyond what the GetText call above gives back.
+                var linePosition = Math.Min(preChangeLineSnapshot.Length, changePosition - preChangeLineSnapshot.Start) - 1;
+
+                if (AfterOpeningBrace(linePosition, preChangeLineSnapshot))
+                {
+                    var afterChangePosition = changePosition + changeLength;
+                    var afterChangeLineSnapshot = currentSnapshot.GetLineFromPosition(afterChangePosition);
+                    var afterChangeLinePosition = afterChangePosition - afterChangeLineSnapshot.Start;
+
+                    if (BeforeClosingBrace(afterChangeLinePosition, afterChangeLineSnapshot))
+                    {
+                        context = new BraceIndentationContext(focusedTextView, changePosition);
+                        return true;
+                    }
+                }
+            }
+
+            context = null;
+            return false;
+        }
+
+        internal static bool BeforeClosingBrace(int linePosition, ITextSnapshotLine lineSnapshot)
+        {
+            var lineText = lineSnapshot.GetText();
+            while (linePosition < lineSnapshot.Length)
+            {
+                if (!char.IsWhiteSpace(lineText[linePosition]))
+                    break;
+                linePosition++;
+            }
+
+            var beforeClosingBrace = linePosition < lineSnapshot.Length && lineText[linePosition] == '}';
+            return beforeClosingBrace;
+        }
+
+        internal static bool AfterOpeningBrace(int linePosition, ITextSnapshotLine lineSnapshot)
+        {
+            var lineText = lineSnapshot.GetText();
+            while (linePosition >= 0)
+            {
+                if (!char.IsWhiteSpace(lineText[linePosition]))
+                    break;
+                linePosition--;
+            }
```

Comment:
This loop might be simpler written as a `for`

Reply:
#taylorcontrolflow